bundles/org.openhab.binding.dwdpollenflug/src/main/java/org/openhab/binding/dwdpollenflug/internal/handler/DWDPollenflugBridgeHandler.java

```diff
@@ -0,0 +1,222 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.dwdpollenflug.internal.handler;
+
+import static org.openhab.binding.dwdpollenflug.internal.DWDPollenflugBindingConstants.*;
+
+import java.net.URI;
+import java.util.List;
+import java.net.SocketTimeoutException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+
+import com.google.gson.Gson;
+import com.google.gson.JsonSyntaxException;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jetty.client.HttpClient;
+import org.eclipse.jetty.client.HttpResponse;
+import org.eclipse.jetty.client.api.Request;
+import org.eclipse.jetty.client.api.Result;
+import org.eclipse.jetty.client.util.BufferingResponseListener;
+import org.eclipse.jetty.http.HttpMethod;
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.binding.BaseBridgeHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.eclipse.smarthome.core.types.RefreshType;
+import org.openhab.binding.dwdpollenflug.internal.DWDPollingException;
+import org.openhab.binding.dwdpollenflug.internal.config.DWDPollenflugBridgeConfiguration;
+import org.openhab.binding.dwdpollenflug.internal.dto.DWDPollenflug;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link DWDPollenflugBridgeHandler} is the handler for bridge thing
+ *
+ * @author Johannes DerOetzi Ott - Initial contribution
+ */
+@NonNullByDefault
+public class DWDPollenflugBridgeHandler extends BaseBridgeHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(DWDPollenflugBridgeHandler.class);
+
+    private static final String DWD_URL = "https://opendata.dwd.de/climate_environment/health/alerts/s31fg.json";
+
+    private DWDPollenflugBridgeConfiguration bridgeConfig = new DWDPollenflugBridgeConfiguration();
+
+    private @Nullable ScheduledFuture<?> pollingJob;
+
+    private @Nullable DWDPollenflug pollenflug;
+
+    private final List<DWDPollenflugRegionListener> regionListeners = new CopyOnWriteArrayList<>();

```

Comment:
I'm not so familiar with Thread-Safe Sets. Can you maybe give me are more detailed example, or a binding I can have a look?

Reply:
Thread-safe implementations can be achived by synchronized collections (i.e. `Collections.synchronizedCollection(...)`. This makes sure that the methods of the collection (like `add`, `remove`, `get`, ...) can only be accessed by one thread at a time, so it is made sure that the same element is never modified by two threads concurrently. The downside is that the whole collection is locked and all othe treads have to wait for the first operazion to complete.

The second option is that you use the collections from `java.util.concurrent`, e.g. the `CopyOnWriteList` you used. This locks only parts of the collections, so that other parts of the collection can still be accessed by other threads. There is no `Set` implementation in this package, but the `ConcurrentHashMap` provides one, so you can use `ConcurrentHashMap.newKeySet()` instead of `new ConcurrentHashMap<>()` which does not exist. 

My second point is that using a `Set` better fits your use-case than a `List`. A `List` can contain duplicates (so if you call `add` several times with the same object, it gets added several times). Each time you iterate over that list, your listener will get called multiple times. Another (more important issue): Calling `remove` will only remove the first element that matches, the duplicates stay in the list. The next time you iterate over the list, one the duplicates are called, even if they should have been removed. 

This can be avoided with a `Set`, because a `Set` only allows distinct objects. If you add the same object a second time, the first one is replaced instead of adding a duplicate. You can even check with the return value (`true` if a new element, `false` if a replaced element) if something strange happened and log a warning in `registerListener`. The same holds true for `unregister`, if you call `remove` and the return-value is `false`, you tried to remove a listener that was not part of the listener collection.

Reply:
Ok I changed now. I had a closer look on it and want to ask:

Wouldn't it be a better pattern not to get the bridge handler inside the region handler and call register and unregister there, but overwrite childHandlerInitialized and childHandlerDisposed in the bridge handler?

Reply:
That's probably also possible. But I think we have two patterns so far in openHAB: registering listeners from the thing to the bridge or (in some cases) the bridge iterating over all things. This would be a third one and I guess there is not much benefit in introducing that.

Reply:
I changed it anyway, Please check again!